src/inmanta/data/__init__.py

```diff
@@ -1187,58 +1194,61 @@ class AgentInstance(BaseDocument):
+    ) -> List[TAgentInstance]:
         if process is not None:
             objects = await cls.get_list(expired=None, tid=tid, name=endpoint, process=process, connection=connection)
         else:
             objects = await cls.get_list(expired=None, tid=tid, name=endpoint, connection=connection)
         return objects
 
     @classmethod
-    async def active(cls):
+    async def active(cls: Type[TAgentInstance]) -> List[TAgentInstance]:
         objects = await cls.get_list(expired=None)
         return objects
 
     @classmethod
-    async def _expire_endpoints_for_process(
-        cls, tid: uuid.UUID, process: uuid.UUID, endpoints: Set[str], now: datetime.datetime
-    ) -> None:
-        """
-        Expire the agent instances for the given endpoints if they are not expired yet.
-        """
-        if not endpoints:
-            return
-        names_subquery = ",".join([f"${i}" for i in range(4, 4 + len(endpoints))])
-        query = f"""UPDATE {cls.table_name()}
-                    SET expired=$1
-                    WHERE tid=$2 AND process=$3 AND name IN ({names_subquery}) AND expired IS NULL
-                 """
-        values = [cls._get_value(now), cls._get_value(tid), cls._get_value(process)] + [cls._get_value(e) for e in endpoints]
-        await cls._execute_query(query, *values)
-
-    @classmethod
-    async def log_instance_creation(
-        cls, tid: uuid.UUID, process: uuid.UUID, endpoints: Set[str], now: datetime.datetime
-    ) -> None:
+    async def log_instance_creation(cls: Type[TAgentInstance], tid: uuid.UUID, process: uuid.UUID, endpoints: Set[str]) -> None:
         """
         Create new agent instances for a given session.
         """
         if not endpoints:
             return
-        # Fix database corruption when database was down
-        await cls._expire_endpoints_for_process(tid, process, endpoints, now)
-        for nh in endpoints:
-            await cls(tid=tid, process=process, name=nh).insert()
+        async with cls.get_connection() as connection:
+            value_tuples: List[Tuple] = [(cls._new_id(), tid, process, name, None) for name in endpoints]
+            value_tuple_len: int = len(value_tuples[0])
+            values_subquery: str = ", ".join(
+                "(%s)" % ", ".join(f"${k}" for k in range(start, start + value_tuple_len))
+                for start in (value_tuple_len * i + 1 for i in range(len(endpoints)))
+            )
+            await connection.execute(
+                f"""
+                INSERT INTO
+                {cls.table_name()}
+                (id, tid, process, name, expired)
+                VALUES {values_subquery}
+                ON CONFLICT ON CONSTRAINT {cls.table_name()}_unique DO UPDATE
+                SET expired = null
+                ;
+                """,
+                *(cls._get_value(v) for v in chain(*value_tuples)),
+            )
+            raise Exception()
```

Comment:
??

Reply:
Don't know how that got there. Must've snuck in somehow after I ran the tests locally because they succeeded. I'll get rid of it of course.